services/server/src/server/services/utils/database-util.ts

```diff
@@ -567,3 +571,64 @@ export function normalizeRecompiledBytecodes(
     });
   }
 }
+
+export async function countSourcifyMatchAddresses(pool: Pool, chain: number) {
+  return await pool.query(
+    `
+    SELECT
+        contract_deployments.chain_id,
+        SUM(CASE WHEN sourcify_matches.creation_match = 'perfect' OR sourcify_matches.runtime_match = 'perfect' THEN 1 ELSE 0 END) AS full_total,
+        SUM(CASE WHEN sourcify_matches.creation_match != 'perfect' AND sourcify_matches.runtime_match != 'perfect' THEN 1 ELSE 0 END) AS partial_total
+    FROM sourcify_matches
+    JOIN verified_contracts ON verified_contracts.id = sourcify_matches.verified_contract_id
+    JOIN contract_deployments ON 
+        contract_deployments.id = verified_contracts.deployment_id
+        AND contract_deployments.chain_id = $1
+    GROUP BY contract_deployments.chain_id;
+    `,
+    [chain]
+  );
```

Comment:
In this case if a user asks a chain's contracts that has couple million addresses, it sends the query. This would take a lot of time no? I thought we were going to set the 200 limit **inside the query** and if the response hits 200 contracts, then we respond on server an error.

Reply:
But this is just for calculating the total, we need a way to let the user know the total of contracts in a chain. Maybe we can optimize this by caching the result

Reply:
Do we need to? Isn't the only important thing if there are more contracts than N? If the DB still needs to count 1M+ rows then it defeats the purpose